identity-account-storage/src/storage/traits.rs

```diff
@@ -27,45 +29,88 @@ mod storage_sub_trait {
   impl<S: Send + Sync + super::Storage> StorageSendSyncMaybe for S {}
 }
 
-/// An interface for Identity Account storage implementations.
+/// An interface for Account storage implementations.
 ///
-/// See [MemStore][crate::storage::MemStore] for a test/example implementation.
+/// The [`Storage`] interface is used for secure operations on keys, such as generation and signing,
+/// as well as key-value like storage of data structures, such as DID documents.
+///
+/// This interface works with [`AccountId`] and [`IotaDID`] as the top-level identifiers.
+/// `AccountId` is intended as a partition-key, i.e. everything related to an account
+/// can be stored in its own partition. Keys belonging to an account are identified by [`KeyLocation`]s
+/// in that partition.
+/// An `IotaDID` can be resolved to an `AccountId` using the index, which is a global data structure.
+/// Therefore, index operations need to be carefully synchronized per [`Storage`] instance.
+/// Other operations don't need to be synchronized globally, as it is a user error to create
+/// more than one `Account` for the same identity.
+///
+/// See [`MemStore`][crate::storage::MemStore] for a test/example implementation.
 #[cfg_attr(not(feature = "send-sync-storage"), async_trait(?Send))]
 #[cfg_attr(feature = "send-sync-storage", async_trait)]
 pub trait Storage: storage_sub_trait::StorageSendSyncMaybe + Debug {
-  /// Write any unsaved changes to disk.
-  async fn flush_changes(&self) -> Result<()>;
-
-  /// Creates a new keypair at the specified `location` and returns its `PublicKey`.
-  async fn key_new(&self, did: &IotaDID, location: &KeyLocation) -> Result<PublicKey>;
-
-  /// Inserts a private key at the specified `location` and returns its `PublicKey`.
-  async fn key_insert(&self, did: &IotaDID, location: &KeyLocation, private_key: PrivateKey) -> Result<PublicKey>;
-
-  /// Retrieves the public key at the specified `location`.
-  async fn key_get(&self, did: &IotaDID, location: &KeyLocation) -> Result<PublicKey>;
-
-  /// Deletes the keypair specified by `location`.
-  async fn key_del(&self, did: &IotaDID, location: &KeyLocation) -> Result<()>;
+  /// Creates a new identity for the given `network`.
+  ///
+  /// - Uses the given Ed25519 `private_key` or generates a new key if it's `None`.
+  /// - Returns an [`IdentityAlreadyExists`](crate::Error::IdentityAlreadyExists) error if the DID already exists.
+  /// - Adds the newly created DID to an index which can be accessed via [`Storage::index`].
+  ///
+  /// Returns the generated DID and the location at which the key was stored.
+  async fn did_create(
```

Comment:
I don't think we can mandate the type of error returned (we don't export error enum variants to bindings as far as I know), just that it returns some sort of error if the identity already exists.

We can throw that error in `Account` ourselves using `did_exists` prior to calling `did_create`.

The entire `Storage` interface doc-comments might need a pass to ensure internal consistency with the expected components. E.g. we mention `an index` here, it should be `the index` or `the DID index` and explain what that is in the doc-comment on `Storage` itself (along with other required components like the private key storage and Document/ChainState storage). Not sure if we should mention an index explicitly since that may just be an implementation detail.

Should we mention the exact functions to use like `KeyLocation::new` and `IotaDID::new`/`IotaDID::new_with_network`?

Reply:
True, as of now, we cannot specify the type we expect. With a potential `StorageError`, we could, as we would probably also export that to bindings, since it just becomes part of the interface at that point.

> just that it returns some sort of error if the identity already exists.

Will do that for now.

> We can throw that error in `Account` ourselves using `did_exists` prior to calling `did_create`.

Can't do that, because the DID is created inside `did_create`, so it's the storage's responsibility to check.

> Not sure if we should mention an index explicitly since that may just be an implementation detail.

Probably not, especially when we rename the `index_*` functions to `did_*`. Mentioning a "list" is probably good enough.

> Should we mention the exact functions

I'm hoping it would be fairly clear, also given example implementations. But I'm mainly trying to save ourselves some work of maintaining docs.

Perhaps a better thing we could do is write a storage test that users can easily plug their storage into, also from bindings, and then we could for example test, that a certain private key of a `KeyPair` returns the expected DID. I realize that would be (much) more work than just documenting it, but also better. Anyway, not trying to do this for this PR.